cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/NativeTwillControllerCreator.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.app.runtime;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.common.twill.TwillAppNames;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import org.apache.twill.api.TwillController;
+import org.apache.twill.api.TwillRunnerService;
+import org.apache.twill.internal.RunIds;
+
+import java.util.Objects;
+
+/**
+ * Implementation of {@link TwillControllerCreator} for On-Premise(Native) Cluster mode.
+ */
+public class NativeTwillControllerCreator implements TwillControllerCreator {
+
+  private final TwillRunnerService twillRunnerService;
+
+  @Inject
+  public NativeTwillControllerCreator(TwillRunnerService twillRunnerService) {
+    this.twillRunnerService = twillRunnerService;
+  }
+
+  @Override
+  public TwillController createTwillController(ProgramRunId programRunId) {
+    return twillRunnerService.lookup(TwillAppNames.toTwillAppName(programRunId.getParent()),
```

Comment:
I'm pretty sure this won't work if you try to start a program in a different kubernetes namespace. 

The reason is that this is relying on watcher threads in KubeTwillRunnerService to add a TwillController to the in-memory liveInfos when it notices that a kubernetes job has been created. However, watcher threads are added lazily for a namespace as part of the prepare() method flow. With all of these changes, that watcher thread will never get added in the KubeTwillRunnerService in app-fabric anymore.

This also has some race condition problems. For example, by the time we look this up in app-fabric, the program could have died, which will cause this to return null.

Reply:
We could potentially do something similar to the Remote case, where it looks up the run record and tells the runner service directly to add a controller, though I'm not sure how feasible that is in the Kube case. For example, one challenge is that the kubernetes namespace is determined when `KubeTwillPreparer.withConfiguration()` is called. It can't just be looked up in the run record detail.